### ISSUE TITLE
Interpreter discovery difference with and without `--python` flag

### ISSUE THREAD

**Issue by matterhorn103** [astral-sh/uv#4277]
Sorry for the spree of issue opening and comment writing, I'm just raising things as I find them. I'm very excited about what you're doing with uv :)

uv version: 0.2.11
OS: Linux

A Python interpreter that is found successfully for a given version string passed with the `--python` flag is not found when the same version is mandated by `requires-python` in a `pyproject.toml`.

It is easiest to demonstrate what I mean, I think:
```bash
~/p/foo> ls
foo.py  pyproject.toml
~/p/foo> cat foo.py
print("Hello World!")

~/p/foo> cat pyproject.toml
[project]
name = "foo"
version = "0.0.1"
requires-python = ">= 3.12"

~/p/foo> python3 -V
Python 3.11.9
~/p/foo> python3.12 -V
Python 3.12.3
~/p/foo> which python3
/usr/bin/python3
~/p/foo> which python3.12
/usr/bin/python3.12
~/p/foo> uv run foo.py
warning: `uv run` is experimental and may change without warning.
error: No interpreter found for Python >=3.12 in search path
~/p/foo [2]> uv run -p 3.12 foo.py
warning: `uv run` is experimental and may change without warning.
Using Python 3.12.3 interpreter at: /usr/bin/python3.12
Creating virtualenv at: .venv
Resolved 1 package in 1ms
   Built foo @ file:///home/matt/python/foo
Downloaded 1 package in 651ms
Installed 1 package in 0.33ms
 + foo==0.0.1 (from file:///home/matt/python/foo)
Hello World!
```

**Comment by zanieb** [comment]
Interesting. Will investigate..

**Comment by zanieb** [comment]
Can you provide verbose logs? `RUST_LOG=uv=trace uv run -v ...`

**Comment by zanieb** [comment]
I get something like

```
❯ RUST_LOG=uv=trace uv run -v -- python -V
warning: `uv run` is experimental and may change without warning.
DEBUG Syncing project environment.
DEBUG Found project root: `/Users/zb/workspace/foo`
DEBUG No workspace root found, using project root
DEBUG Searching for Python >=3.12 in search path
TRACE Searching PATH for executables: python3, python
TRACE Checking `PATH` directory for interpreters: /Users/zb/bin
TRACE Checking `PATH` directory for interpreters: /Users/zb/.local/bin
TRACE Checking `PATH` directory for interpreters: /Users/zb/.cargo/bin
TRACE Checking `PATH` directory for interpreters: /opt/homebrew/bin
TRACE Found possible Python executable: /opt/homebrew/bin/python3
TRACE Cached interpreter info for Python 3.12.3, skipping probing: /opt/homebrew/bin/python3
DEBUG Found CPython 3.12.3 at `/opt/homebrew/bin/python3` (search path)
Using Python 3.12.3 interpreter at: /opt/homebrew/opt/python@3.12/bin/python3.12
Creating virtualenv at: .venv
...
DEBUG Finished building: foo @ file:///Users/zb/workspace/foo
Downloaded 1 package in 419ms
Installed 1 package in 0.75ms
 + foo==0.0.1 (from file:///Users/zb/workspace/foo)
DEBUG Running `python -V`
Python 3.12.3
```

and the same if I include the `--python` flag.

**Comment by zanieb** [comment]
Ah so the difference here is that unless you _request_ Python 3.12 we won't look for an interpreter with the `3.12` suffix. The key difference is

```
TRACE Searching PATH for executables: python3.12, python3, python
```

vs

```
TRACE Searching PATH for executables: python3, python
```

If you were to pass the flag `--python '>=3.12'` you'd get the same result. We _could_ consider searching for executables with suffixes for Python range requests (i.e. >=3.12), it just gets kind of complicated since that range is technically unbounded.

**Comment by zanieb** [comment]
I was thinking about adding this but am actually unsure where we'd stop. Like should we be checking for `python3.33` on the PATH? Probably not.

**Comment by matterhorn103** [comment]
I think at minimum the lowest possible matching version should be checked for, as that can be pretty much guaranteed to be a released version. I.e. `--python '>=3.12'`should check `python3.12, python3, python` and `--python '>3.9'` should check `python3.10, python3, python` (I assume using > is possible there but I don't actually know).

When it comes to scanning across many possible version numbers, shouldn't it be possible to come up with a reasonable upper bound based on the current date? The Python release schedule is fairly predictable after all, so in the event of no upper bound being given for the Python version, can't an upper bound be set to `YY - 11 + x`, where `x` is a buffer, something like 3, for safety?

**Comment by konstin** [comment]
We need to split `PATH` at the path separator, list all directories and filter them for `python3\.([\d]+)` where the minor number is compatible.

**Comment by zanieb** [comment]
There's `which_re` though a regular expression sounds kind of heavy here it would be a good start.

**Comment by zanieb** [comment]
I'm going to close this in favor of https://github.com/astral-sh/uv/issues/4709 which specifies a way we could improve this behavior.

At the moment though, this is working as designed.

**Comment by zanieb** [comment]
@matterhorn103 this is resolved in #5148
